core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java

```diff
@@ -1164,6 +1168,192 @@ public UnifyResult apply(UnifyRuleCall call) {
+      for (int i = 0; i < query.rowType.getFieldCount(); i++) {
+        if (i < fieldCnt(qInput0)) {
+          compenProjs.add(
+              new RexInputRef(i, query.rowType.getFieldList().get(i).getType()));
+        } else {
+          final RexNode shifted = RexUtil.shift(qInput1Projs.get(i - fieldCnt(qInput0)),
+              qInput0.rowType.getFieldCount());
+          compenProjs.add(shifted);
+        }
+      }
+      final RexProgram compenRexProgram = RexProgram.create(
+          target.rowType, compenProjs, compenCond,
+          query.rowType, rexBuilder);
+      final MutableCalc compenCalc = MutableCalc.of(target, compenRexProgram);
+
+      return tryMergeParentCalcAndGenResult(call, compenCalc);
+    }
+  }
+
+  /**
+   * A {@link SubstitutionVisitor.UnifyRule} that matches a {@link MutableCorrelate} which has
+   * {@link MutableCalc} as children to a {@link MutableCorrelate}. We try to pull up the {@link
+   * MutableCalc} to top of {@link MutableCorrelate}, then match the {@link MutableCorrelate} in
+   * query to {@link MutableCorrelate} in target.
+   */
+  private static class CorrelateOnCalcsToCorrelateUnifyRule extends AbstractUnifyRule {
+
+    public static final CorrelateOnCalcsToCorrelateUnifyRule INSTANCE =
+        new CorrelateOnCalcsToCorrelateUnifyRule();
+
+    private CorrelateOnCalcsToCorrelateUnifyRule() {
+      super(
+          operand(MutableCorrelate.class,
+              operand(MutableCalc.class, query(0)), operand(MutableCalc.class, query(1))),
+          operand(MutableCorrelate.class, target(0), target(1)), 2);
+    }
+
+    @Override protected UnifyResult apply(UnifyRuleCall call) {
+      final MutableCorrelate query = (MutableCorrelate) call.query;
+      final MutableCalc qInput0 = (MutableCalc) query.getLeft();
+      final MutableCalc qInput1 = (MutableCalc) query.getRight();
+      final Pair<RexNode, List<RexNode>> qInput0Explained = explainCalc(qInput0);
+      final RexNode qInput0Cond = qInput0Explained.left;
+      final List<RexNode> qInput0Projs = qInput0Explained.right;
+      final Pair<RexNode, List<RexNode>> qInput1Explained = explainCalc(qInput1);
+      final RexNode qInput1Cond = qInput1Explained.left;
+      final List<RexNode> qInput1Projs = qInput1Explained.right;
+
+      final MutableCorrelate target = (MutableCorrelate) call.target;
+
+      final RexBuilder rexBuilder = call.getCluster().getRexBuilder();
+      final JoinRelType joinRelType = sameJoinType(query.joinType, target.joinType);
+      if (joinRelType == null) {
+        return null;
+      }
+      if (joinRelType != JoinRelType.INNER && !(joinRelType.isOuterJoin()
+          && qInput0Cond.isAlwaysTrue() && qInput1Cond.isAlwaysTrue())) {
+        return null;
+      }
+      if (!checkRequiredColumnsAndCorrelationId(query, target)) {
+        return null;
+      }
+
+      final RexNode qInput1CondShifted =
+          RexUtil.shift(qInput1Cond, fieldCnt(qInput0.getInput()));
```

Comment:
What if correlate variable is referenced in condition, should we pull it up ?

Reply:
i think we should pull it up.